test/test_matmul_scheduler.cpp

```diff
@@ -369,15 +369,15 @@ TEST_P(PrecisionParametrizedTest, EpilogueBiasRelu) {
 //   D = A x B;
 //   Aux = relu(D)
 //  Target architectures: Turing, Ampere
-TEST_P(PrecisionParametrizedTest, DISABLED_EpilogueReluAux) {
+TEST_P(PrecisionParametrizedTest, EpilogueReluAux) {
```

Comment:
**Note:** the epilogue has not be merged and there are two loops with the same ranges, one executed after another. The very similar test case, `EpilogueReluAux`, produces epilogue with inlined loops.

Sources of kernel generated for A100 for HSH test case:
```
__global__ void nvfuser_matmul_f0_c1_r0_g0(Tensor<__half, 2, 2> T0, Tensor<__half, 2, 2> T1, Tensor<__half, 2, 2> T6, Tensor<__half, 2, 2> T8) {
  alignas(16) extern __shared__ char array[];
  const unsigned smem_offset = 0;
  nvfuser_index_t i0 = ceilDiv(T0.logical_size[1LL], 32);
  nvfuser_index_t i1 = 8 * T1.logical_size[1LL];
  nvfuser_index_t i2 = ((nvfuser_index_t)threadIdx.x) / 4;
  nvfuser_index_t i3 = ((nvfuser_index_t)threadIdx.x) % 4;
  nvfuser_index_t i4 = 8LL * i3;
  nvfuser_index_t i5 = T1.logical_size[1LL] * i2;
  nvfuser_index_t i6 = i1 * ((nvfuser_index_t)threadIdx.y);
  nvfuser_index_t i7 = i1 * ((nvfuser_index_t)threadIdx.z);
  nvfuser_index_t i8 = 32LL * ((nvfuser_index_t)blockIdx.y);
  __half* ptr9 = ((((T1.data + i4) + i5) + i6) + i7) + i8;
  nvfuser_index_t i10 = 32LL * T1.logical_size[1LL];
  nvfuser_index_t i11 = 64 * i2;
  nvfuser_index_t i12 = i3 / 4;
  nvfuser_index_t i13 = 64 * i12;
  nvfuser_index_t i14 = i3 % 4;
  nvfuser_index_t i15 = 16 * (((i2 / 2) % 4) ^ i14);
  nvfuser_index_t i16 = 512 * ((nvfuser_index_t)threadIdx.y);
  nvfuser_index_t i17 = 512 * ((nvfuser_index_t)threadIdx.z);
  nvfuser_index_t i18 = 8 * T0.logical_size[1LL];
  nvfuser_index_t i19 = T0.logical_size[1LL] * i2;
  nvfuser_index_t i20 = i18 * ((nvfuser_index_t)threadIdx.y);
  nvfuser_index_t i21 = i18 * ((nvfuser_index_t)threadIdx.z);
  nvfuser_index_t i22 = (256LL * T0.logical_size[1LL]) * ((nvfuser_index_t)blockIdx.x);
  __half* ptr23 = ((((T0.data + i4) + i19) + i20) + i21) + i22;
  nvfuser_index_t i24 = 32 * T0.logical_size[1LL];
  nvfuser_index_t i25 = ((nvfuser_index_t)threadIdx.x) / 8;
  nvfuser_index_t i26 = 512 * (i25 % 2);
  nvfuser_index_t i27 = ((nvfuser_index_t)threadIdx.x) % 8;
  nvfuser_index_t i28 = 64 * i27;
  nvfuser_index_t i29 = i25 / 2;
  nvfuser_index_t i30 = 64 * (i29 / 4);
  nvfuser_index_t i31 = (i27 / 2) % 4;
  nvfuser_index_t i32 = 16 * (i31 ^ (i29 % 4));
  nvfuser_index_t i33 = 4096 * ((nvfuser_index_t)threadIdx.z);
  nvfuser_index_t i34 = ((nvfuser_index_t)threadIdx.x) % 16;
  nvfuser_index_t i35 = 64 * i34;
  nvfuser_index_t i36 = ((nvfuser_index_t)threadIdx.x) / 16;
  nvfuser_index_t i37 = 8 * (i36 % 2);
  nvfuser_index_t i38 = 64 * ((nvfuser_index_t)threadIdx.y);
  nvfuser_index_t i39 = i36 / 2;
  nvfuser_index_t i40 = (i34 / 2) % 4;
  __half* ptr41 = (((((T1.data + (64 * T1.logical_size[1LL])) + i4) + i5) + i6) + i7) + i8;
  __half* ptr42 = (((((64 + T0.data) + i4) + i19) + i20) + i21) + i22;
  nvfuser_index_t i43 = 2LL + i29;
  nvfuser_index_t i44 = 64 * (i43 / 4);
  nvfuser_index_t i45 = 16 * (i31 ^ (i43 % 4));
  nvfuser_index_t i46 = (((32LL * i2) + (2 * i14)) + (32LL * ((nvfuser_index_t)threadIdx.y))) + (2048LL * ((nvfuser_index_t)threadIdx.z));
  nvfuser_index_t i47 = i2 % 4;
  nvfuser_index_t i48 = (256 * ((nvfuser_index_t)threadIdx.y)) + (256 * ((nvfuser_index_t)threadIdx.z));
  nvfuser_index_t i49 = 8LL * ((nvfuser_index_t)threadIdx.x);
  nvfuser_index_t i50 = 8 * i3;
  nvfuser_index_t i51 = ((((i50 + i5) + i6) + i7) + i8) + ((256LL * T1.logical_size[1LL]) * ((nvfuser_index_t)blockIdx.x));
  nvfuser_index_t i52 = 32 * T1.logical_size[1LL];
  bool b53 = ((7 + i50) + i8) < T1.logical_size[1LL];
  nvfuser_index_t i54 = 8 * ((nvfuser_index_t)threadIdx.y);
  nvfuser_index_t i55 = 8 * ((nvfuser_index_t)threadIdx.z);
  nvfuser_index_t i56 = (((-T0.logical_size[1LL]) + i2) + i54) + i55;
  nvfuser_index_t i57 = (7 - T0.logical_size[1LL]) + i50;
  nvfuser_index_t i58 = ((((-T0.logical_size[0LL]) + i2) + i54) + i55) + (256LL * ((nvfuser_index_t)blockIdx.x));
  nvfuser_index_t i59 = ((64 + i2) + i54) + i55;
  nvfuser_index_t i60 = 71 + i50;
  __half* T13 = reinterpret_cast<__half*>(array + smem_offset + 81920LL);
  __half* T12 = reinterpret_cast<__half*>(array + smem_offset + 32768LL);
  float* T11 = reinterpret_cast<float*>(array + smem_offset + 0);
  unsigned i61 = toSmem(T13);
  unsigned i62 = ((((i61 + i11) + i13) + i15) + i16) + i17;
  unsigned i63 = toSmem(T12);
  unsigned i64 = ((((i63 + i11) + i13) + i15) + i16) + i17;
  unsigned i65 = (i63 + i26) + i28;
  unsigned i66 = ((i65 + i30) + i32) + i33;
  unsigned i67 = ((i61 + i35) + i37) + i38;
  unsigned i68 = ((i65 + i44) + i45) + i33;
  unsigned i69 = (((1024 + i61) + i35) + i37) + i38;
  Array<float, 64, 2> T5;
  #pragma unroll
  for(nvfuser_index_t i70 = 0; i70 < 4; ++i70) {
    nvfuser_index_t i71 = 16 * i70;
    #pragma unroll
    for(nvfuser_index_t i72 = 0; i72 < 4; ++i72) {
      Ampere::initM16N8K16(*reinterpret_cast<Array<float, 4, 1>*>(&T5[(i71 + (4 * i72))]));
    }
  }
  #pragma unroll
  for(nvfuser_index_t i73 = 0; i73 < 2; ++i73) {
    nvfuser_index_t i74 = 32LL * i73;
    __half* ptr75 = ptr23 + i74;
    unsigned i76 = i64 + (16384 * i73);
    nvfuser_index_t i77 = -i74;
    bool b78 = i57 < i77;
    Ampere::cpAsyncCg<__half, 8>((i62 + (2048 * i73)), (ptr9 + (i10 * i73)), (b53 && (i56 < i77)));
    #pragma unroll
    for(nvfuser_index_t i79 = 0; i79 < 8; ++i79) {
      Ampere::cpAsyncCg<__half, 8>((i76 + (2048 * i79)), (ptr75 + (i24 * i79)), (b78 && (i58 < (-(32 * i79)))));
    }
    asm volatile("cp.async.commit_group;\n");
  }
  asm volatile("cp.async.wait_group %0;\n"::"n"(1LL));
  __syncthreads();
  Array<__half, 64, 8> T14;
  Array<__half, 32, 4> T2;
  #pragma unroll
  for(nvfuser_index_t i80 = 0; i80 < 4; ++i80) {
    Turing::ldMatrix(*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T14[(8LL * i80)]), (i66 + (1024 * i80)));
  }
  #pragma unroll
  for(nvfuser_index_t i81 = 0; i81 < 4; ++i81) {
    nvfuser_index_t i82 = i39 + i81;
    Turing::ldMatrixT(*reinterpret_cast<Array<uint32_t, 2, 1>*>(&T2[(4LL * i81)]), ((i67 + (64 * (i82 / 4))) + (16 * (i40 ^ (i82 % 4)))));
  }
  #pragma unroll 1
  for(nvfuser_index_t i83 = 0; i83 < i0; ++i83) {
    nvfuser_index_t i84 = (2 + i83) % 3;
    nvfuser_index_t i85 = 32LL * i83;
    __half* ptr86 = ptr42 + i85;
    unsigned i87 = i64 + (16384 * i84);
    nvfuser_index_t i88 = i83 % 3;
    unsigned i89 = i68 + (16384 * i88);
    unsigned i90 = i69 + (2048 * i88);
    nvfuser_index_t i91 = (1 + i83) % 3;
    unsigned i92 = i66 + (16384 * i91);
    unsigned i93 = i67 + (2048 * i91);
    bool b94 = (i60 + i85) < T0.logical_size[1LL];
    Ampere::cpAsyncCg<__half, 8>((i62 + (2048 * i84)), (ptr41 + (i10 * i83)), (b53 && ((i59 + i85) < T0.logical_size[1LL])));
    #pragma unroll
    for(nvfuser_index_t i79 = 0; i79 < 8; ++i79) {
      Ampere::cpAsyncCg<__half, 8>((i87 + (2048 * i79)), (ptr86 + (i24 * i79)), (b94 && (i58 < (-(32 * i79)))));
    }
    asm volatile("cp.async.commit_group;\n");
    #pragma unroll
    for(nvfuser_index_t i80 = 0; i80 < 4; ++i80) {
      Turing::ldMatrix(*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T14[(32 + (8LL * i80))]), (i89 + (1024 * i80)));
    }
    __half T3[32];
    #pragma unroll
    for(nvfuser_index_t i95 = 0; i95 < 4; ++i95) {
      nvfuser_index_t i96 = 8LL * i95;
      #pragma unroll
      for(nvfuser_index_t i97 = 0; i97 < 8; ++i97) {
        T3[(i96 + i97)] = 0.00000000000000000e+00;
      }
    }
    #pragma unroll
    for(nvfuser_index_t i95 = 0; i95 < 4; ++i95) {
      nvfuser_index_t i98 = 8LL * i95;
      #pragma unroll
      for(nvfuser_index_t i97 = 0; i97 < 8; ++i97) {
        nvfuser_index_t i99 = i98 + i97;
        T3[i99] = T14[i99];
      }
    }
    #pragma unroll
    for(nvfuser_index_t i81 = 0; i81 < 4; ++i81) {
      nvfuser_index_t i100 = i39 + i81;
      Turing::ldMatrixT(*reinterpret_cast<Array<uint32_t, 2, 1>*>(&T2[(16 + (4LL * i81))]), ((i90 + (64 * (i100 / 4))) + (16 * (i40 ^ (i100 % 4)))));
    }
    __half T4[16];
    #pragma unroll
    for(nvfuser_index_t i101 = 0; i101 < 4; ++i101) {
      nvfuser_index_t i102 = 4LL * i101;
      #pragma unroll
      for(nvfuser_index_t i103 = 0; i103 < 4; ++i103) {
        T4[(i102 + i103)] = 0.00000000000000000e+00;
      }
    }
    #pragma unroll
    for(nvfuser_index_t i101 = 0; i101 < 4; ++i101) {
      nvfuser_index_t i104 = 4LL * i101;
      #pragma unroll
      for(nvfuser_index_t i103 = 0; i103 < 4; ++i103) {
        nvfuser_index_t i105 = i104 + i103;
        T4[i105] = T2[i105];
      }
    }
    #pragma unroll
    for(nvfuser_index_t i70 = 0; i70 < 4; ++i70) {
      nvfuser_index_t i106 = 16 * i70;
      #pragma unroll
      for(nvfuser_index_t i72 = 0; i72 < 4; ++i72) {
        Ampere::M16N8K16TNF16(
          *reinterpret_cast<Array<float, 4, 1>*>(&T5[(i106 + (4 * i72))]),
          (reinterpret_cast<Array<unsigned,4, 1>*>(&T3)[i70]),
          (reinterpret_cast<Array<unsigned,2, 1>*>(&T4)[i72]));
      }
    }
    #pragma unroll
    for(nvfuser_index_t i95 = 0; i95 < 4; ++i95) {
      nvfuser_index_t i107 = 8LL * i95;
      #pragma unroll
      for(nvfuser_index_t i97 = 0; i97 < 8; ++i97) {
        T3[(i107 + i97)] = 0.00000000000000000e+00;
      }
    }
    #pragma unroll
    for(nvfuser_index_t i95 = 0; i95 < 4; ++i95) {
      nvfuser_index_t i108 = 8LL * i95;
      nvfuser_index_t i109 = 32 + i108;
      #pragma unroll
      for(nvfuser_index_t i97 = 0; i97 < 8; ++i97) {
        T3[(i108 + i97)] = T14[(i109 + i97)];
      }
    }
    #pragma unroll
    for(nvfuser_index_t i101 = 0; i101 < 4; ++i101) {
      nvfuser_index_t i110 = 4LL * i101;
      #pragma unroll
      for(nvfuser_index_t i103 = 0; i103 < 4; ++i103) {
        T4[(i110 + i103)] = 0.00000000000000000e+00;
      }
    }
    #pragma unroll
    for(nvfuser_index_t i101 = 0; i101 < 4; ++i101) {
      nvfuser_index_t i111 = 4LL * i101;
      nvfuser_index_t i112 = 16 + i111;
      #pragma unroll
      for(nvfuser_index_t i103 = 0; i103 < 4; ++i103) {
        T4[(i111 + i103)] = T2[(i112 + i103)];
      }
    }
    #pragma unroll
    for(nvfuser_index_t i70 = 0; i70 < 4; ++i70) {
      nvfuser_index_t i113 = 16 * i70;
      #pragma unroll
      for(nvfuser_index_t i72 = 0; i72 < 4; ++i72) {
        Ampere::M16N8K16TNF16(
          *reinterpret_cast<Array<float, 4, 1>*>(&T5[(i113 + (4 * i72))]),
          (reinterpret_cast<Array<unsigned,4, 1>*>(&T3)[i70]),
          (reinterpret_cast<Array<unsigned,2, 1>*>(&T4)[i72]));
      }
    }
    asm volatile("cp.async.wait_group %0;\n"::"n"(1LL));
    __syncthreads();
    #pragma unroll
    for(nvfuser_index_t i80 = 0; i80 < 4; ++i80) {
      Turing::ldMatrix(*reinterpret_cast<Array<uint32_t, 4, 1>*>(&T14[(8LL * i80)]), (i92 + (1024 * i80)));
    }
    #pragma unroll
    for(nvfuser_index_t i81 = 0; i81 < 4; ++i81) {
      nvfuser_index_t i114 = i39 + i81;
      Turing::ldMatrixT(*reinterpret_cast<Array<uint32_t, 2, 1>*>(&T2[(4LL * i81)]), ((i93 + (64 * (i114 / 4))) + (16 * (i40 ^ (i114 % 4)))));
    }
  }
  #pragma unroll
  for(nvfuser_index_t i115 = 0; i115 < 4; ++i115) {
    nvfuser_index_t i116 = 16 * i115;
    nvfuser_index_t i117 = i46 + (512LL * i115);
    #pragma unroll
    for(nvfuser_index_t i118 = 0; i118 < 4; ++i118) {
      nvfuser_index_t i119 = i116 + (4 * i118);
      nvfuser_index_t i120 = i12 + i118;
      nvfuser_index_t i121 = (i117 + (32LL * (i120 / 4))) + (8LL * (i47 ^ (i120 % 4)));
      #pragma unroll
      for(nvfuser_index_t i122 = 0; i122 < 2; ++i122) {
        loadGeneric<float, 2>( &T11[(i121 + (256LL * i122))],  &T5[(i119 + (2LL * i122))]);
      }
    }
  }
  __syncthreads();
  #pragma unroll
  for(nvfuser_index_t i123 = 0; i123 < 8; ++i123) {
    nvfuser_index_t i124 = i48 + (1024 * i123);
    Array<__half, 8, 8> T9;
    #pragma unroll
    for(nvfuser_index_t i125 = 0; i125 < 8; ++i125) {
      nvfuser_index_t i126 = i49 + i125;
      nvfuser_index_t i127 = i126 / 32;
      nvfuser_index_t i128 = i126 % 32;
      nvfuser_index_t i129 = i128 / 8;
      T9[i125] = __float2half(T11[((((i124 + (32LL * i127)) + (32LL * (i129 / 4))) + (8LL * ((i127 % 4) ^ (i129 % 4)))) + (i128 % 8))]);
    }
    if ((b53 && (i58 < (-(32 * i123))))) {
      loadLocalToGlobal<__half, /*vec_size=*/8, /*is_volatile=*/false>( &T6[(i51 + (i52 * i123))], &T9[0]);
    }
  }
  #pragma unroll
  for(nvfuser_index_t i130 = 0; i130 < 8; ++i130) {
    nvfuser_index_t i131 = i48 + (1024 * i130);
    Array<__half, 8, 8> T10;
    #pragma unroll
    for(nvfuser_index_t i132 = 0; i132 < 8; ++i132) {
      nvfuser_index_t i133 = i49 + i132;
      nvfuser_index_t i134 = i133 / 32;
      nvfuser_index_t i135 = i133 % 32;
      nvfuser_index_t i136 = i135 / 8;
      float T7[1];
      T7[0] = relu(T11[((((i131 + (32LL * i134)) + (32LL * (i136 / 4))) + (8LL * ((i134 % 4) ^ (i136 % 4)))) + (i135 % 8))]);
      T10[i132] = __float2half(T7[0]);
    }
    if ((b53 && (i58 < (-(32 * i130))))) {
      loadLocalToGlobal<__half, /*vec_size=*/8, /*is_volatile=*/false>( &T8[(i51 + (i52 * i130))], &T10[0]);
    }
  }
}
```

Reply:
> **Note:** the epilogue has not be merged and there are two loops with the same ranges, one executed after another. The very similar test case, `EpilogueReluAux`, produces epilogue with inlined loops.

If I understand correctly, "merging" these two loops are horizontal merges (something like merging `T1[4] = T0[4]; T3[4] = T2[4]` into a single loop), rather than vertical merges (something like merging `T1[4] = T0[4]; T2[4] = T1[4]` into a single loop)? If so, merging them do not sound like something each to schedule with the current infrastructure. I think we should just create an issue for this saying that we need to revisit this when we start to do more detailed perf tuning, and focus on functionality for now.

Reply:
I have created an issue (https://github.com/NVIDIA/Fuser/issues/1360) as advised.